### PR TITLE
ArrayScopeSymbol use dyncast() to recognize the array content

### PR DESCRIPTION
**Pull request by ghost**
Among the 3 fields that give the array content, only one is set at a time.
By chance the type of each field has different dyncasts, so take advantage of this
to save a small block of memory.

when building dparse as a static library I observe a reduction of the memory consumed by 800 Kb. (from ~905Mb to ~904Mb)

### CODE REVIEW

File: src/dmd/dsymbol.h [hunk 1]
```diff
@@ -330,9 +330,7 @@ class WithScopeSymbol : public ScopeDsymbol
 class ArrayScopeSymbol : public ScopeDsymbol
 {
 public:
-    Expression *exp;    // IndexExp or SliceExp
-    TypeTuple *type;    // for tuple[length]
-    TupleDeclaration *td;       // for tuples of objects
+    RootObject *scopedArray;
```
Comment (thewilsonator) [inline src/dmd/dsymbol.h:333]: this is private in D, but this will need to be visible to GDC & LDC

Comment (NilsLankila) [reply src/dmd/dsymbol.h:333]: no, it's rather that I forgot to set it private in c++

Comment (NilsLankila) [reply src/dmd/dsymbol.h]: updated

**Comment by NilsLankila** [comment]
added a second commit to get rid of a label, better review commit by commit now.

**Review by ibuclaw — APPROVED** [review]

**Pull request merged**